File: app/src/main/java/com/mrboomdev/awery/ui/activity/SplashActivity.java
package com.mrboomdev.awery.ui.activity;

import static com.mrboomdev.awery.app.App.enableEdgeToEdge;
import static com.mrboomdev.awery.app.App.getDatabase;
import static com.mrboomdev.awery.app.App.resolveAttrColor;
import static com.mrboomdev.awery.app.CrashHandler.reportIfCrashHappened;
import static com.mrboomdev.awery.util.async.AsyncUtils.thread;

import android.annotation.SuppressLint;
import android.content.Intent;
import android.os.Bundle;
import android.util.Log;

import androidx.annotation.Nullable;
import androidx.appcompat.app.AppCompatActivity;
import androidx.core.splashscreen.SplashScreen;

import com.mrboomdev.awery.app.AweryLifecycle;
import com.mrboomdev.awery.app.CrashHandler;
import com.mrboomdev.awery.databinding.ScreenSplashBinding;
import com.mrboomdev.awery.extensions.ExtensionsFactory;
import com.mrboomdev.awery.generated.AwerySettings;
import com.mrboomdev.awery.ui.ThemeManager;
import com.mrboomdev.awery.ui.activity.setup.SetupActivity;
import com.mrboomdev.awery.util.async.AsyncFuture;

@SuppressLint("CustomSplashScreen")
public class SplashActivity extends AppCompatActivity {
	private static final String TAG = "SplashActivity";
	private ScreenSplashBinding binding;

	@Override
	protected void onCreate(@Nullable Bundle savedInstanceState) {
		SplashScreen.installSplashScreen(this);

		try {
			ThemeManager.apply(this);
		} catch(Exception e) {
			Log.e(TAG, "Failed to apply an theme!", e);
		}

		enableEdgeToEdge(this);
		super.onCreate(savedInstanceState);

		binding = ScreenSplashBinding.inflate(getLayoutInflater());
		binding.getRoot().setBackgroundColor(resolveAttrColor(this, android.R.attr.colorBackground));
		getWindow().setNavigationBarColor(resolveAttrColor(this, android.R.attr.colorBackground));
		setContentView(binding.getRoot());

		binding.status.setText("Checking the database...");

		reportIfCrashHappened(this, () -> thread(() -> {
			try {
				getDatabase().getListDao().getAll();
			} catch(IllegalStateException e) {
				Log.e(TAG, "Database is corrupted!", e);

				CrashHandler.showErrorDialog(this, new CrashHandler.CrashReport.Builder()
						.setTitle("Database is corrupted!")
						.setThrowable(e)
						.setDismissCallback(AweryLifecycle::exitApp)
						.build());

				return;
			}

			if(AwerySettings.SETUP_VERSION_FINISHED.getValue() < SetupActivity.SETUP_VERSION) {
				var intent = new Intent(this, SetupActivity.class);
				startActivity(intent);
				finish();
				return;
			}

			ExtensionsFactory.getInstance().addCallback(new AsyncFuture.Callback<>() {
				@Override
				public void onSuccess(ExtensionsFactory result) {
					startActivity(new Intent(SplashActivity.this, MainActivity.class));
					finish();
				}

				@Override
				public void onFailure(Throwable t) {
					Log.e(TAG, "Failed to load an ExtensionsFactory!", t);

					CrashHandler.showErrorDialog(SplashActivity.this, new CrashHandler.CrashReport.Builder()
							.setTitle("Failed to load an ExtensionsFactory")
							.setThrowable(t)
							.setDismissCallback(AweryLifecycle::exitApp)
							.build());
				}
			});

			runOnUiThread(this::update);
		}));
	}

	private void update() {
		if(isDestroyed()) return;
		var factory = ExtensionsFactory.getInstanceNow();

		if(factory == null) {
			binding.status.setText("Loading extensions...");
			return;
		}

		long progress = 0, total = 0;

		for(var manager : factory.getManagers()) {
			var managerProgress = manager.getProgress();
			progress += managerProgress.getProgress();
			total += managerProgress.getMax();
		}

		binding.status.setText("Loading extensions " + progress + "/" + total);
		AweryLifecycle.runDelayed(this::update, 100);
	}
}

File: app/src/main/java/com/mrboomdev/awery/ui/activity/SplashActivity.kt
package com.mrboomdev.awery.ui.activity

import android.annotation.SuppressLint
import android.os.Bundle
import android.util.Log
import androidx.appcompat.app.AppCompatActivity
import androidx.core.splashscreen.SplashScreen.Companion.installSplashScreen
import androidx.lifecycle.lifecycleScope
import com.mrboomdev.awery.BuildConfig
import com.mrboomdev.awery.R
import com.mrboomdev.awery.app.App.Companion.database
import com.mrboomdev.awery.app.App.Companion.isTv
import com.mrboomdev.awery.app.AweryLifecycle.Companion.exitApp
import com.mrboomdev.awery.app.AweryLifecycle.Companion.runDelayed
import com.mrboomdev.awery.app.CrashHandler
import com.mrboomdev.awery.app.ExtensionsManager
import com.mrboomdev.awery.databinding.ScreenSplashBinding
import com.mrboomdev.awery.extensions.ExtensionsFactory
import com.mrboomdev.awery.generated.AwerySettings
import com.mrboomdev.awery.ui.activity.settings.setup.SetupActivity
import com.mrboomdev.awery.ui.tv.TvMainActivity
import com.mrboomdev.awery.util.async.AsyncFuture
import com.mrboomdev.awery.util.extensions.applyTheme
import com.mrboomdev.awery.util.extensions.enableEdgeToEdge
import com.mrboomdev.awery.util.extensions.resolveAttrColor
import com.mrboomdev.awery.util.extensions.startActivity
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.flow.catch
import kotlinx.coroutines.flow.collect
import kotlinx.coroutines.flow.onCompletion
import kotlinx.coroutines.flow.onEach
import kotlinx.coroutines.launch

private const val USE_NEW_SOURCES = false

@SuppressLint("CustomSplashScreen")
class SplashActivity : AppCompatActivity() {
	private lateinit var binding: ScreenSplashBinding

	override fun onCreate(savedInstanceState: Bundle?) {
		installSplashScreen()

		try {
			applyTheme()
		} catch(e: Exception) {
			Log.e(TAG, "Failed to apply an theme!", e)
		}

		enableEdgeToEdge()
		super.onCreate(savedInstanceState)

		binding = ScreenSplashBinding.inflate(layoutInflater).apply {
			root.setBackgroundColor(resolveAttrColor(android.R.attr.colorBackground))
			status.setText(R.string.checking_if_crash_occurred)
		}

		window.navigationBarColor = resolveAttrColor(android.R.attr.colorBackground)
		setContentView(binding.root)

		CrashHandler.showDialogIfCrashHappened(this) {
			binding.status.setText(R.string.checking_database)

			lifecycleScope.launch(Dispatchers.IO) {
				try {
					database.listDao.all
				} catch(e: IllegalStateException) {
					Log.e(TAG, "Database is corrupted!", e)

					CrashHandler.showDialog(
						context = this@SplashActivity,
						titleRes = R.string.database_corrupted,
						throwable = e,
						dismissCallback = ::exitApp)

					return@launch
				}

				if(AwerySettings.SETUP_VERSION_FINISHED.value < SetupActivity.SETUP_VERSION) {
					startActivity(SetupActivity::class)
					finish()
					return@launch
				}

				if(USE_NEW_SOURCES) {
					ExtensionsManager.init(this@SplashActivity).onEach {
						launch(Dispatchers.Main) {
							binding.status.text = getString(R.string.loading_extensions_n, it.progress, it.max)
						}
					}.onCompletion {
						// Tv version isn't done yet at 100%
						startActivity(if(isTv && BuildConfig.DEBUG) TvMainActivity::class else MainActivity::class)
						finish()
					}.catch {
						CrashHandler.showDialog(
							context = this@SplashActivity,
							title = "Failed to load an ExtensionsFactory",
							throwable = it,
							dismissCallback = ::exitApp)
					}.collect()

					return@launch
				}

				ExtensionsFactory.getInstance().addCallback(object : AsyncFuture.Callback<ExtensionsFactory?> {
					override fun onSuccess(result: ExtensionsFactory) {
						// Tv version isn't done yet at 100%
						startActivity(if(isTv && BuildConfig.DEBUG) TvMainActivity::class else MainActivity::class)
						finish()
					}

					override fun onFailure(t: Throwable) {
						Log.e(TAG, "Failed to load an ExtensionsFactory!", t)

						CrashHandler.showDialog(
							context = this@SplashActivity,
							title = "Failed to load an ExtensionsFactory",
							throwable = t,
							dismissCallback = ::exitApp)
					}
				})

				runOnUiThread { update() }
			}
		}
	}

	private fun update() {
		if(isDestroyed) return
		val factory = ExtensionsFactory.getInstanceNow()

		if(factory == null) {
			binding.status.setText(R.string.loading_extensions)
			return
		}

		var progress: Long = 0
		var total: Long = 0

		for(manager in factory.managers) {
			val managerProgress = manager.progress
			progress += managerProgress.progress
			total += managerProgress.max
		}

		binding.status.text = getString(R.string.loading_extensions_n, progress, total)
		runDelayed({ this.update() }, 100)
	}

	companion object {
		private const val TAG = "SplashActivity"
	}
}